### PR TITLE
Service loadable executor prototype

### PR DESCRIPTION
**Pull request by hutchig**
Motivation:

My goal is to open a code based discussion for an approach that might address #3434. 
This have a Vert.x SPI that provides the ability to use a foreign ExecutorService
for worker tasks.

This is to enable easier consumption of SmallRye technologies that make use of Vert.x in
environments with custom ExecutorServices (in the first instance OpenLiberty).

This is a draft PR for now and I am very happy to make any further improvements or changes that might make
this change more likely to be adopted either while it is a draft or when it is turned into a 'standard' PR.

### CODE REVIEW

**Comment by vietj** [comment]
can we have this PR onto master instead of 3.9 ?

**Comment by hutchig** [comment]
Hi @vietj yes certainly I will do a PR onto master directly now.

I did it to the 3.9 line initially because of https://github.com/smallrye/smallrye-reactive-messaging/blob/0225686abdc52dbec1253d8ae8c29531bcf70d3a/pom.xml#L60
which is the dependency for the specific use case I was tasked to solve.

Thank you very much for commenting on this.

**Comment by vietj** [comment]
can we review this PR @hutchig ?

**Comment by hutchig** [comment]
Yes, that would be great @vietj I went through the CI tests this morning and none of the failures seem to be related to this PR as far as I can see. Thank you. As always, happy to change anything at all or add tests, doc etc.

**Comment by vietj** [comment]
I reran CI one the two build failed

**Comment by vietj** [comment]
I think we could simplify a bit the implementation

1. currently the implementation uses `new ThreadPoolExecutor(..)` and `Executors.newFixedThreadPool` which seem to be similar

2. always use `ExecutorServiceFactory` to create a thread pool, so when no `ExecutorServiceFactory` is provided by the SPI then we use an internal implementation that returns a fixed thread pool.

**Comment by hutchig** [comment]
The two different forms were there in the code already and I did not want to alter the existing behaviour - I will make it as you suggest above, thanks for the feedback @vietj .

**Comment by hutchig** [comment]
@vietj For "always use ExecutorServiceFactory to create a thread pool," I can use my head and not apply that to transport.eventLoopGroup but will take what you say and apply it to the internalBlockingExec.

**Comment by vietj** [comment]
@hutchig yes indeed that's the point

File: src/test/java/io/vertx/test/core/LogContainsRule.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+/**
+ * A generic log scanner used by testcases to check that messages have appeared
+ * in the JUnit test logs.
+ *
+ */
+public class LogContainsRule implements TestRule {
```
Comment (vietj) [inline src/test/java/io/vertx/test/core/LogContainsRule.java:35]: we have a similar rule already existing : `io.vertx.test.core.BlockedThreadWarning`, can you reuse this class and perhaps make it more generic to it works in this case too ?

File: src/test/java/io/vertx/core/spi/executor/ExecutorFactoryMetricsTest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * 
+ * It performs the same tests as MetricsContextTest but using an external
+ * {@link java.util.concurrent.ExecutorService} for worker tasks
+ * 
+ * To enable different tests to have particular ExecutorServiceFactory's (or
+ * none), we only put the particular ExecutorServiceFactory .class file on the
+ * classpath and make it service loadable as part of each test and are careful
+ * to remove it afterwards.
+ */
+public class ExecutorFactoryMetricsTest extends MetricsContextTest {
```
Comment (vietj) [inline src/test/java/io/vertx/core/spi/executor/ExecutorFactoryMetricsTest.java:38]: I don't think we need this as it tests metrics behavior.

File: src/test/java/io/vertx/core/spi/executor/Utils.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+import org.junit.rules.ExternalResource;
+import org.junit.rules.RuleChain;
+import org.junit.rules.TestRule;
+
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.test.core.LogContainsRule;
+
+public class Utils {
```
Comment (vietj) [inline src/test/java/io/vertx/core/spi/executor/Utils.java:39]: ServiceHelperTest is using also Java compiler to do something similar, can you update the ServiceHelperTest to use this class instead?

File: src/test/java/io/vertx/test/core/LogContainsRule.java (hunk 1 above)
Comment (hutchig) [reply src/test/java/io/vertx/test/core/LogContainsRule.java:35]: Yes, will do.

File: src/test/java/io/vertx/core/spi/executor/ExecutorFactoryMetricsTest.java (hunk 2 above)
Comment (hutchig) [reply src/test/java/io/vertx/core/spi/executor/ExecutorFactoryMetricsTest.java:38]: Agreed.

File: src/test/java/io/vertx/core/spi/executor/Utils.java (hunk 3 above)
Comment (hutchig) [reply src/test/java/io/vertx/core/spi/executor/Utils.java:39]: Yes OK will do, probably tomorrow now.

**Comment by vietj** [comment]
in master I removed this compilation code and replaced with maven compiler
execution:
https://github.com/eclipse-vertx/vert.x/commit/82e0c4d690146c59c62439697ddba575b281c4e0

File: src/test/java/io/vertx/test/core/LogContainsRule.java (hunk 1 above)
Comment (hutchig) [reply src/test/java/io/vertx/test/core/LogContainsRule.java:36]: Done. (Hope @tsegismont is OK with the author tag being left as I learned how to do the superclass from this one.)

File: src/test/java/io/vertx/core/spi/executor/ExecutorFactoryMetricsTest.java (hunk 2 above)
Comment (hutchig) [reply src/test/java/io/vertx/core/spi/executor/ExecutorFactoryMetricsTest.java]: Done.

File: src/test/java/io/vertx/core/spi/executor/Utils.java (hunk 3 above)
Comment (hutchig) [reply src/test/java/io/vertx/core/spi/executor/Utils.java:40]: Having looked at 82e0c4d more closely, does the above still stand or are you going to backport
what master does?

Comment (hutchig) [reply src/test/java/io/vertx/core/spi/executor/Utils.java:40]: I chose that it was part of the test to move SPI service classes on and off the classpath in the tests to try to make them more realistic - that is why I did not have them compiled and on the classpath all at once. However, its true that the ServiceLoader provider configuration files alone could be used and it is not our job to test Java ServiceLoading. I copied the style and code from one that Clement wrote.

@vietj What would you like me to do:

1. backport 82e0c4d and eliminate the compilation from the tests in my PR
2. as mentioned above, make 3.9 ServiceHelper and my PR tests use a common Utils method (maybe relocated out of executor/Utils to a more generic location)
3. nothing, as you are going to backport 82e0c4d at some point
?

**Comment by hutchig** [comment]
I was asked to check with @cescoffier (with an '@' here) if he knew of any reason why this would be incompatible with the `@Blocking` in SmallRye Reactive messaging.

**Comment by cescoffier** [comment]
@hutchig it should be fine.

File: src/test/java/io/vertx/core/spi/executor/Utils.java (hunk 3 above)
Comment (vietj) [reply src/test/java/io/vertx/core/spi/executor/Utils.java:40]: option 1. is my favorite, it's a relevant cleanup

Comment (hutchig) [reply src/test/java/io/vertx/core/spi/executor/Utils.java:40]: That's fine by me. I will learn more as it was doing mnv stuff I have not done that way before
and I agree it is more efficient not to compile 'in' the test.
I will push another commit on Monday.

**Comment by hutchig** [comment]
CI Test failed on:
```
Failed tests: 
  EventBusRegistrationRaceTest.theTest:82->AsyncTestBase.assertEquals:235 expected:<0> but was:<1>
Tests run: 4287, Failures: 1, Errors: 0, Skipped: 14
```

File: pom.xml [hunk 4]
```diff
@@ -571,6 +571,62 @@
...
+              <goal>testCompile</goal>
+            </goals>
+            <configuration>
+              <compileSourceRoots>
+                <compileSourceRoot>src/test/classpath/servicehelper</compileSourceRoot>
+              </compileSourceRoots>
+              <outputDirectory>${project.build.directory}/classpath/servicehelper</outputDirectory>
+            </configuration>
+          </execution>
+          <execution>
```
Comment (hutchig) [inline pom.xml:591]: The reason that there is a second clause here is that there is a pattern in the test code of creating a new classloader and adding a directory such as the  _${project.build.directory}/classpath/servicehelper_ above to it's classpath. If I had a single folder, all the test classloaders would have combined classpaths and in order to seperate them, the smallest chunk of XML that I can duplicate is a an <execution>.

File: src/test/java/io/vertx/core/spi/executor/Utils.java (hunk 3 above)
Comment (hutchig) [reply src/test/java/io/vertx/core/spi/executor/Utils.java:29]: Done.

**Comment by vietj** [comment]
is it ready for another review @hutchig

**Comment by hutchig** [comment]
Yes please, that would be great @vietj thank you.

File: src/main/java/io/vertx/core/impl/VertxImpl.java [hunk 5]
```diff
@@ -1082,6 +1087,13 @@ void release() {
     }
   }
 
+  public class DefaultExecutorServiceFactory implements ExecutorServiceFactory {
```
Comment (vietj) [inline src/main/java/io/vertx/core/impl/VertxImpl.java:1090]: can we instead use `FixedThreadPoolExecutorServiceFactory` that seems to be the same class ?

File: src/test/java/io/vertx/core/spi/executor/Utils.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+   * methods
+   */
+  static Method EXECUTE = null;
+  static Method SHUTDOWN_NOW = null;
+  static {
+    try {
+      EXECUTE = ExecutorService.class.getMethod("execute", Runnable.class);
+      SHUTDOWN_NOW = ExecutorService.class.getMethod("shutdownNow");
+    } catch (NoSuchMethodException | SecurityException e) {
+      throw new IllegalStateException(e);
```
Comment (vietj) [inline src/test/java/io/vertx/core/spi/executor/Utils.java:54]: I think that will prevent the class loading instead we should set null fields and fail in the other methods with an assertion that check the fields are not null before using them or throwing an IllegalStateException at this place

File: src/main/java/io/vertx/core/impl/VertxImpl.java (hunk 5 above)
Comment (vietj) [reply src/main/java/io/vertx/core/impl/VertxImpl.java:1090]: my bad, FixedThreadPoolExecutorServiceFactory is in test package, forget it.

**Comment by vietj** [comment]
beside the minor static initializer throwing IllegalStateException it seems fine to me.

File: src/test/java/io/vertx/core/spi/executor/Utils.java (hunk 6 above)
Comment (hutchig) [reply src/test/java/io/vertx/core/spi/executor/Utils.java:54]: Ok I will make it so. Originally the methods parameters were not typed, only string names but what you suggest is better now they are the actual method types.

Comment (hutchig) [reply src/test/java/io/vertx/core/spi/executor/Utils.java]: Done, is ready for another review @vietj I think. Thank you again.

**Comment by hutchig** [comment]
fixes #3434

**Comment by hutchig** [comment]
I could hug you, thank you.

**Comment by vietj** [comment]
I will port this to master with the new service provider

**Pull request merged**